pkg/gatewayserver/io/mqtt/format_protobufv2.go

```diff
@@ -112,7 +112,7 @@ func (protobufv2) ToUplink(message []byte, ids *ttnpb.GatewayIdentifiers) (*ttnp
 			return nil, err
 		}
 		settings.DataRate = loraDr.DataRate
-		settings.CodingRate = lorawanMetadata.CodingRate
+		settings.DataRate.Modulation.(*ttnpb.DataRate_Lora).Lora.CodingRate = lorawanMetadata.CodingRate
```

Comment:
I would move this inside the switch, just to make sure that `settings.DataRate.Modulation` really has type `*ttnpb.DataRate_Lora` (and you can edit `loraDr` directly).

Reply:
If I'm not mistaken it is already positioned within the switch statement, so the type is set at that point. As `loraDr` is also of type `DataRate` we would still need to imply a type on it to edit it. 

Might be a nicer solution to adjust the `ParseLoRa` function to also take a coding rate. But this would cause an issue with the `UnmarshallJSON` function which has no information about the coding rate.